test/common/grpc/google_async_client_cache_test.cc

```diff
@@ -0,0 +1,98 @@
+                                 const ::envoy::config::core::v3::GrpcService config) {
+    AsyncClientCacheSharedPtr actual_client_cache =
+        client_cache_singleton_->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_,
+                                                             config);
+    EXPECT_EQ(expected_client_cache, actual_client_cache);
+    EXPECT_EQ(expected_client, actual_client_cache->getAsyncClient());
+  }
+
+  void expectCacheAndClientNotEqual(const AsyncClientCacheSharedPtr& expected_client_cache,
+                                    const RawAsyncClientSharedPtr& expected_client,
+                                    const ::envoy::config::core::v3::GrpcService config) {
+    AsyncClientCacheSharedPtr actual_client_cache =
+        client_cache_singleton_->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_,
+                                                             config);
+    EXPECT_NE(expected_client_cache, actual_client_cache);
+    EXPECT_NE(expected_client, actual_client_cache->getAsyncClient());
+  }
+
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  Grpc::MockAsyncClientManager async_client_manager_;
+  Grpc::MockAsyncClient* async_client_ = nullptr;
+  Grpc::MockAsyncClientFactory* factory_ = nullptr;
+  NiceMock<Stats::MockIsolatedStatsStore> scope_;
+  std::unique_ptr<AsyncClientCacheSingleton> client_cache_singleton_;
+  std::unique_ptr<AsyncClientCache> client_cache_;
+};
+
+TEST_F(AsyncClientCacheTest, Deduplication) {
+  Stats::IsolatedStoreImpl scope;
+  testing::InSequence s;
+
+  ::envoy::config::core::v3::GrpcService config;
+  config.mutable_google_grpc()->set_target_uri("dns://test01");
+  config.mutable_google_grpc()->set_credentials_factory_name("test_credential01");
+
+  expectClientCreation();
+  AsyncClientCacheSharedPtr test_client_cache_01 =
+      client_cache_singleton_->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_,
+                                                           config);
+  RawAsyncClientSharedPtr test_client_01 = test_client_cache_01->getAsyncClient();
+  // Fetches the existing client and they should be equal.
+  expectCacheAndClientEqual(test_client_cache_01, test_client_01, config);
+
+  config.mutable_google_grpc()->set_credentials_factory_name("test_credential02");
+  expectClientCreation();
+  // Different credentials use different clients.
+  expectCacheAndClientNotEqual(test_client_cache_01, test_client_01, config);
+
+  AsyncClientCacheSharedPtr test_client_cache_02 =
+      client_cache_singleton_->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_,
+                                                           config);
+  RawAsyncClientSharedPtr test_client_02 = test_client_cache_02->getAsyncClient();
+
+  config.mutable_google_grpc()->set_credentials_factory_name("test_credential02");
+  // No creation, fetching the existing one.
+  expectCacheAndClientEqual(test_client_cache_02, test_client_02, config);
```

Comment:
Thanks for doing this. It's more readable and clear now. One slight nit is that if there's a failure it will be reported in a line number in the helper function and you won't know which call is at fault.

You could remedy by returning an AssertionResult or in a few other ways, but it's not mission-critical to the PR obviously.

Reply:
Done. Added better error message